### PR TITLE
mrc-3885: Add endpoint for uploading input file

### PR DESCRIPTION
**Pull request by r-ash**
This builds on #441 

This PR will
* Add an endpoint which will allow an input file to be uploaded

### CODE REVIEW

File: R/upload.R [hunk 1]
```diff
@@ -0,0 +1,18 @@
+upload_input <- function(queue) {
+  function(file, filename) {
+    md5 <- toupper(digest::digest(file, algo = "md5"))
+    dest_path <- file.path(
+      queue$uploads_dir,
+      paste0(md5, ".", tools::file_ext(filename)))
+    if (!file.exists(dest_path)) {
+      t <- tempfile()
+      writeBin(file, t)
+      file.copy(t, dest_path)
```
Comment (r-ash) [inline R/upload.R:10]: This will be safe with multiple hintr instances running?

Comment (richfitz) [reply R/upload.R:10]: it depends what you're trying to do. If you want an atomic write, then you need to make the tempfile be on the same filesystem as the destination, then use `file.rename`. So if `dest_path` is your upload directory, we could do

```
dest_path <- file.path(
  queue$uploads_dir,
  paste0(md5, ".", tools::file_ext(filename)))
if (!file.exists(dest_path)) {
  tmpdir <- file.path(queue$uploads_dir, ".incoming"))
  dir.create(tmpdir, FALSE, TRUE
  t <- tempfile(tmpdir = tmpdir)
  writeBin(file, t)
  file.rename(t, dest_path)
}
```

as you have it, you might as well do `writeBin(file, dest_path)` as that will be at least as fast as the later `file.copy`

Of course,  if the other copies of hint/hintr are doing a plain write you have a potential race condition, but tbh it's not that likely

File: R/api.R [hunk 2]
```diff
@@ -366,6 +367,18 @@ endpoint_adr_metadata <- function(queue) {
                                     returning = response)
 }
 
+
+endpoint_upload_input <- function(queue) {
+  input <- porcelain::porcelain_input_body_binary("file")
+  response <- porcelain::porcelain_returning_json(
+    "SessionFile.schema", schema_root())
+  porcelain::porcelain_endpoint$new("POST",
+                                    "/upload/input/<filename>",
```
Comment (richfitz) [inline R/api.R:376]: I wonder about putting this below some other top level path (internal/upload perhaps) if there are many to make it clear it's not really part of the public API?

File: R/upload.R (hunk 1 above)
Comment (r-ash) [reply R/upload.R]: Thanks, though not likely would be good to avoid it. Have added this, thanks!

File: R/api.R (hunk 2 above)
Comment (r-ash) [reply R/api.R]: Yeah sounds like a good idea

**Review by richfitz — APPROVED** [review]

**Pull request merged**